keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,489 @@
+In addition, to make things worse, provisioner secrets that are defined by using template with PVC information,
+like "${pvc.name}", would be deleted in above case, due to the failure in finding a reference to the secret,
+because the reference to provisioner secret is not stored in PV and needs to be resolved on deletion time with PVC information that is already deleted.
+This behavior is inconsistent with other secrets, like node publish secret whose reference is stored in PV.
+
+## Design Details
+
+This feature can be implemented in the same way as `pv-protection/pvc-protection`.
+Protection logics for in-tree resources and out-of-tree resources are separeted and independently work.
+It is due to the restriction that CRDs can't be handled from in-tree controller.
+
+- In-tree resources(`Pod` and `PersistentVolume`):
+  - The deletion is blocked by using newly introduced `kubernetes.io/secret-protection` finalizer,
+  - The `kubernetes.io/secret-protection` finalizer will be always added on creation of the secret by using admission controller for in-tree resources,
+  - After the secret is requested to be deleted (`deletionTimestamp` is set), the `kubernetes.io/secret-protection` finalizer will be deleted by newly introduced in-tree `secret-protection-controller` by checking whether the secret is in-use, on every change(Create/Update/Delete) events for secrets and related resources.
+  - If the `SecretInUseProtection` feature gate is disalbed, finalizer is added on creation, but deleted regardless of whether it is in-use after `deletionTimestamp` is set. This will allow users to avoid manually deleting finalizers on the downgrading by disabling the feature.
+- Out-of-tree resources(`VolumeSnapshot`):
+  - The deletion is blocked by using newly introduced `snapshot.storage.kubernetes.io/secret-protection` finalizer,
+  - The `snapshot.storage.kubernetes.io/secret-protection` finalizer will be always added on creation of the secret by using admission controller for `VolumeSnapshot`,
+  - After the secret is requested to be deleted (`deletionTimestamp` is set), the `snapshot.storage.kubernetes.io/secret-protection` finalizer will be deleted by newly introduced out-of-tree `secret-protection-volumesnapshot-controller` by checking whether the secret is in-use, on every change(Create/Update/Delete) events for secrets and related resources.
+
+Feature gate `SecretInUseProtection` is used only for in-tree controller. Out-of-tree controller will be always enabled when the `secret-protection-volumesnapshot-controller` is deployed.
+
+For users to force delete the secret, users need to do either:
+1. manually delete the finalizer, by below command:
+    ```
+    kubectl patch secret/secret-to-be-deleted -p '{"metadata":{"finalizers":[]}}' --type=merge
+    ```
+2. add `secret.kubernetes.io/disable-protection: "yes"` annotation to opt-out this feature per secret
```

Comment:
So user can't really force delete if the disable protection annotation is not added.  Instead of calling this "force delete", can we call this "skip secret protection"?

Reply:
User can either delete the finalizer or add the annotation. "skip secret protection" sounds good.